automatic/nexus-repository/tools/chocolateyinstall.ps1

```diff
@@ -35,18 +42,89 @@ if (Test-Path "$env:ProgramFiles\nexus\bin") {
 }
 
 if ((Get-Service $ServiceName -ErrorAction SilentlyContinue)) {
-  Write-Warning "Nexus web app is already present, shutting it down so that we can upgrade it."
+  $CurrentlyInstalledVersion = Get-NexusVersion
+  Write-Warning "Nexus web app $($CurrentlyInstalledVersion) is already present, shutting it down so that we can upgrade it."
   Get-Service $ServiceName | Stop-Service -Force
 }
 
-if ($pp.ContainsKey("BackupSslConfig")) {
+if ($pp.ContainsKey("BackupSslConfig") -or $CurrentConfig.'application-port-ssl' -gt 0) {
   if ($pp.ContainsKey("BackupLocation")) {
     Backup-NexusSSL -BackupLocation $pp["BackupLocation"]
   } else {
     Backup-NexusSSL
   }
 }
 
+if ($CurrentlyInstalledVersion -and [version]($CurrentlyInstalledVersion -replace '-\d+$') -lt "3.70") {
+  Write-Error (@(
+    "Please upgrade nexus-repository to version 3.70.1-02 before upgrading further."
+    "You can do this by running 'choco upgrade nexus-repository --version 3.70.1.2 --confirm'"
+    "For more details, please see: https://help.sonatype.com/en/orient-pre-3-70-java-8-or-11.html"
+  ) -join "`n")
+  throw "Package cannot upgrade from '$($CurrentlyInstalledVersion)' to '$($env:ChocolateyPackageVersion)'"
+} elseif ($CurrentlyInstalledVersion -and [version]($CurrentlyInstalledVersion -replace '-\d+$') -eq "3.70.1" -and (Test-NexusDatabaseType -Type "OrientDb" -DataDir $TargetDataFolder -ProgramDir $TargetFolder)) {
+  Write-Host "Preparing for database migration..."
+  $MigrationFiles = Join-Path $ExtractFolder "dbmigration"
+  if ((Get-PSDrive $MigrationFiles[0]).Free -lt ($RequiredSpace = [Math]::Max((3 * (Get-ChildItem -Path $TargetDataFolder\nexus3\db -Recurse | Measure-Object Length -Sum).Sum), 10GB))) {
+    Write-Error -Message (@(
+      "The Sonatype Nexus Database Migrator requires at least $([Math]::Round(($RequiredSpace / 1GB), 2))GB free space."
+      "There is only $([Math]::Floor((Get-PSDrive $MigrationFiles[0]).Free / 1GB))GB free. For more details, please see: "
+      "https://help.sonatype.com/en/orient-3-70-java-8-or-11.html#considerations-before-migrating-252166"
+    ) -join "`n")
+    throw "Cannot migrate database with available disk space"
+  }
+  if (($Memory = Get-CimInstance Win32_PhysicalMemory).Capacity -lt 16GB) {
+    Write-Warning -Message (@(
+      "The Sonatype Nexus Database Migrator requires at least 16GB free space."
+      "There is only $($Memory.Capacity/1GB)GB available. For more details, please see: "
+      "https://help.sonatype.com/en/orient-3-70-java-8-or-11.html#considerations-before-migrating-252166"
+    ) -join "`n")
+  }
+  New-NexusOrientDbBackup -DataDir $TargetDataFolder -ServiceName $ServiceName -DestinationPath $MigrationFiles
+
+  # See: https://help.sonatype.com/en/orientdb-downloads.html
+  $MigratorDownload = @{
+    PackageName = $env:ChocolateyPackageName
+    Url = 'https://download.sonatype.com/nexus/nxrm3-migrator/nexus-db-migrator-3.70.1-03.jar'
```

Comment:
Is it worth breaking this migrator out into a separate package so it can be updated separately, since the download page mentions that download urls for older version(s) will be removed if/when it gets updated.

Reply:
I'm going back and forth between "we should continue to support migration for several versions, so this makes good sense" (as we'd have to consider adding the migration tool to the update script in either case) and "unless we plan to support migrating indefinitely, and will consequently have to continue to upgrade this tool and test mirgations, maybe we should just say 'the upgrade path requires you to go to 3.70.1, then 3.71.x, and then it's all good'".

![image](https://github.com/user-attachments/assets/a7761921-629f-4c17-92b8-a8363f27006d)

I dislike this broken chain.